src/ck-core/ckcheckpoint.C

```diff
@@ -214,14 +220,25 @@ void CkCheckpointMgr::Checkpoint(const char *dirname, CkCallback cb, bool _reque
 	requestStatus = _requestStatus;
 	// make dir on all PEs in case it is a local directory
 	CmiMkdir(dirname);
-	bool success = true;
-        if (CmiNumPartitions() > 1) {
-          ostringstream partDir;
-          partDir << dirname;
-          addPartitionDirectory(partDir);
-          CmiMkdir(partDir.str().c_str());
-        }
 
+  // Create partition directories (if applicable)
+  ostringstream dirPath;
+  dirPath << dirname;
+  if (CmiNumPartitions() > 1) {
+    addPartitionDirectory(dirPath);
+    CmiMkdir(dirPath.str().c_str());
+  }
+
+  // Create subdirectories by grouping PEs
+  // Nodegroups are fine since # of logical nodes <= # of PEs
+  int numPes = CkNumPes();
+  int numPeGroups = ceil((double)numPes / PE_GROUP_SIZE);
+  for (int i = 0; i < numPeGroups; i++) {
+    dirPath << "/sub" << i;
```

Comment:
Won't this keep making dirPath longer and longer for every subgroup? (e.g. `/sub0sub1sub2...`)

Reply:
The above will fix this problem.